crates/lang_srv/src/registry.rs

```diff
@@ -1,75 +1,193 @@
+    TextEdit, Url,
 };
 
-use crate::analysis::{AnalyzedDocument, GlobalAnalysis};
+use crate::analysis::{AnalyzedDocument, DocInfo};
+
+#[derive(Debug)]
+pub(crate) struct DocumentPair {
+    info: DocInfo,
+    latest_document: OnceLock<Arc<AnalyzedDocument>>,
+    last_good_document: Arc<AnalyzedDocument>,
+}
 
-pub(crate) enum DocumentChange {
-    Modified(Url, String),
-    Closed(Url),
+impl DocumentPair {
+    pub(crate) fn new(
+        latest_doc: Arc<AnalyzedDocument>,
+        last_good_document: Arc<AnalyzedDocument>,
+    ) -> Self {
+        Self {
+            info: latest_doc.doc_info.clone(),
+            latest_document: OnceLock::from(latest_doc),
+            last_good_document,
+        }
+    }
 }
 
 #[derive(Debug, Default)]
 pub(crate) struct Registry {
-    documents: HashMap<Url, AnalyzedDocument>,
+    documents: Mutex<HashMap<Url, DocumentPair>>,
 }
 
 impl Registry {
-    pub fn apply_change(&mut self, change: DocumentChange) {
-        match change {
-            DocumentChange::Modified(url, source) => {
-                let GlobalAnalysis { documents } = GlobalAnalysis::new(url, source);
-
-                // Only replace the set of documents and all dependencies that were re-analyzed.
-                // Note that this is actually the opposite of what we want - in truth we want to
-                // re-evaluate all dependents!
-                for document in documents {
-                    let url = document.url().clone();
-                    self.documents.insert(url.clone(), document);
+    pub async fn get_latest_version(&self, url: &Url) -> Option<i32> {
+        self.documents
+            .lock()
+            .await
+            .get(&url)
+            .map(|x| x.info.version)
+    }
+
+    fn update_document<'a>(
+        documents: &mut MutexGuard<'a, HashMap<Url, DocumentPair>>,
+        document: Arc<AnalyzedDocument>,
+        updating_url: &Url,
+    ) {
+        if &document.doc_info.url == updating_url {
+            documents
+                .get_mut(&updating_url)
+                .map(|a| a.latest_document.set(document.clone()).unwrap());
+        }
+
+        let url = document.url().clone();
+        match documents.get_mut(&url) {
+            Some(old_doc) => {
+                if document.type_checked() {
+                    *old_doc = DocumentPair::new(document.clone(), document);
+                } else {
+                    debug!(
+                        "Document typechecking failed at version {:?}, not updating last_good_document",
+                        &document.doc_info.version
+                    );
+                    *old_doc = DocumentPair::new(document, old_doc.last_good_document.clone());
                 }
             }
-            DocumentChange::Closed(_url) => {
-                // Do nothing.
+            None => {
+                documents.insert(url.clone(), DocumentPair::new(document.clone(), document));
+            }
+        }
```

Comment:
I would still bail out here if the `document` we're updating with has a version number that is stale. Otherwise, we might update a stale document that was changed before the lock was acquired.